src/core/ddsc/src/dds_qos_provider.c

```diff
@@ -0,0 +1,290 @@
+        dds_qos_kind_t kind;
+        switch(qos->kind)
+        {
+          case DDS_SYSDEF_TOPIC_QOS:
+            kind = DDS_TOPIC_QOS;
+            break;
+          case DDS_SYSDEF_READER_QOS:
+            kind = DDS_READER_QOS;
+            break;
+          case DDS_SYSDEF_WRITER_QOS:
+            kind = DDS_WRITER_QOS;
+            break;
+          case DDS_SYSDEF_SUBSCRIBER_QOS:
+            kind = DDS_SUBSCRIBER_QOS;
+            break;
+          case DDS_SYSDEF_PUBLISHER_QOS:
+            kind = DDS_PUBLISHER_QOS;
+            break;
+          case DDS_SYSDEF_PARTICIPANT_QOS:
+            kind = DDS_PARTICIPANT_QOS;
+            break;
+          default:
+            ddsrt_free(prefix);
+            ddsrt_free(item);
+            goto err_prov;
+        }
+        item->kind = kind;
+        item->qos = dds_create_qos();
+        dds_merge_qos(item->qos, qos->qos);
+        if (qos->name != NULL)
+          (void) ddsrt_asprintf(&item->full_name, "%s"PROVIDER_ITEM_SEP"%s", prefix, qos->name);
+        else
+          item->full_name = ddsrt_strdup(prefix);
+        if (!ddsrt_hh_add(keyed_qos, item))
+        {
+          QOSPROV_ERROR("Qos duplicate name: %s kind: %d file: %s.\n",
+                        item->full_name, item->kind, path);
+          ret = DDS_RETCODE_BAD_PARAMETER;
+          ddsrt_free(prefix);
+          cleanup_qos_items(item, NULL);
+          goto err_prov;
+        }
+      }
+      ddsrt_free(prefix);
+    }
+  }
+  qos_provider->file_path = ddsrt_strdup(path);
+  qos_provider->keyed_qos = keyed_qos;
+  *provider = qos_provider;
+
+  return ret;
+err_prov:
+  ddsrt_hh_enum(keyed_qos, cleanup_qos_items, NULL);
+  ddsrt_hh_free(keyed_qos);
+  ddsrt_free(qos_provider);
+  return ret;
+}
+
+dds_return_t dds_create_qos_provider (const char *path, dds_qos_provider_t **provider)
+{
+  dds_return_t ret = DDS_RETCODE_OK;
+  if ((ret = dds_create_qos_provider_scope (path, provider, PROVIDER_ITEM_SCOPE_NONE)) != DDS_RETCODE_OK)
+    goto err;
+
+err:
+  return ret;
+}
+
+dds_return_t dds_qos_provider_get_qos (const dds_qos_provider_t *provider, dds_qos_kind_t type, const char *key, dds_qos_t **qos)
+{
+  dds_return_t ret = DDS_RETCODE_OK;
+  if (provider == NULL || provider->keyed_qos == NULL)
+  {
+    QOSPROV_WARN("Failed to access provider qos\n");
+    ret = DDS_RETCODE_BAD_PARAMETER;
+    goto err;
+  }
+  QOSPROV_TRACE("request qos for entity type: %d, scope: %s", type, key);
+  dds_qos_item_t it = {.full_name = ddsrt_strdup(key), .kind = type};
+  dds_qos_item_t *item;
+  if ((item = ddsrt_hh_lookup(provider->keyed_qos, &it)) == NULL)
+  {
+    QOSPROV_WARN("Failed to get qos with name: %s, kind: %d ref file: %s\n",
+                            it.full_name, it.kind, provider->file_path);
+    ret = DDS_RETCODE_BAD_PARAMETER;
+    goto err2;
+  }
+  *qos = item->qos;
+
+err2:
+  ddsrt_free(it.full_name);
+err:
+  return ret;
+}
+
+#define FILL_TOKEN_STR(start,end,token) \
```

Comment:
I don't see why this (and `FILL_TOKENS_STR`, `EMPTY_TOKEN_STR` and `EMPTY_TOKENS_STR`) has to be a macro, and that means it is probably better done as a function.